// @midnight-ntwrk/wallet-sdk-abstractions/dist/InMemoryTransactionHistoryStorage.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { Schema } from 'effect';
/**
 * In-memory implementation of the TransactionHistoryStorage interface.
 *
 * An optional `merge` function can be provided to control how existing and incoming entries are combined during
 * {@link upsert}. When omitted the default behaviour is a shallow spread (`{ ...existing, ...incoming }`).
 *
 * Because the merge runs **synchronously** inside `upsert`, the single-threaded nature of JavaScript guarantees
 * atomicity — no external semaphore is needed.
 */
export class InMemoryTransactionHistoryStorage {
    storage;
    schema;
    merge;
    constructor(schema, merge) {
        this.storage = new Map();
        this.schema = schema;
        this.merge = merge ?? ((existing, incoming) => ({ ...existing, ...incoming }));
    }
    upsert(entry) {
        const existing = this.storage.get(entry.hash);
        this.storage.set(entry.hash, existing ? this.merge(existing, entry) : entry);
        return Promise.resolve();
    }
    getAll() {
        return Array.fromAsync(this.storage.values());
    }
    get(hash) {
        return Promise.resolve(this.storage.get(hash));
    }
    reset() {
        this.storage.clear();
    }
    async serialize() {
        const allEntries = await this.getAll();
        const encode = Schema.encodeSync(Schema.Array(this.schema));
        return JSON.stringify(encode([...allEntries]));
    }
    static restore(serialized, schema, merge) {
        const decode = Schema.decodeUnknownSync(Schema.Array(schema));
        const decoded = decode(JSON.parse(serialized));
        const storage = new InMemoryTransactionHistoryStorage(schema, merge);
        for (const entry of decoded) {
            storage.storage.set(entry.hash, entry);
        }
        return storage;
    }
}
