// @midnight-ntwrk/wallet-sdk-abstractions/dist/TransactionHistoryStorage.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { Schema } from 'effect';
export const TransactionHashSchema = Schema.String;
export const TransactionHistoryStatusSchema = Schema.Literal('SUCCESS', 'FAILURE', 'PARTIAL_SUCCESS');
export const TransactionHistoryCommonSchema = Schema.Struct({
    hash: TransactionHashSchema,
    protocolVersion: Schema.Number,
    status: TransactionHistoryStatusSchema,
    identifiers: Schema.optional(Schema.Array(Schema.String)),
    timestamp: Schema.optional(Schema.Date),
    fees: Schema.optional(Schema.NullOr(Schema.BigInt)),
});
